File: iso_compliance/public/js/purchase_order.js
// Supplier approval standing shown as soon as a supplier is picked on a
// Purchase Order. Enforcement is server-side (overrides/purchase_order.py).

function hcc_show_supplier_standing(frm) {
	if (!frm.doc.supplier || frm.doc.docstatus !== 0) return;
	frappe.db
		.get_value("Supplier", frm.doc.supplier, [
			"custom_approval_status",
			"custom_score",
			"custom_reapproval_due",
		])
		.then((r) => {
			const s = r && r.message;
			if (!s) return;
			frm.set_intro("");
			const status = s.custom_approval_status;
			const score = s.custom_score ? ` — ${(s.custom_score / 20).toFixed(1)}★` : "";
			if (status === "Approved") {
				frm.set_intro(__("Supplier is Approved{0} (REG-007).", [score]), "green");
			} else if (status === "Suspended") {
				frm.set_intro(__("Supplier is Suspended — this order cannot be submitted (SOP-005)."), "red");
			} else {
				frm.set_intro(
					__("Supplier is {0} on the Approved Suppliers Register — an FRM-005 evaluation is expected before regular orders (SOP-005).", [
						status ? __(status) : __("not yet classified"),
					]),
					"orange"
				);
			}
		});
}

frappe.ui.form.on("Purchase Order", {
	refresh: hcc_show_supplier_standing,
	supplier: hcc_show_supplier_standing,
});

File: iso_compliance/overrides/purchase_order.py
# Copyright (c) 2026, Hatim Carbon Co. Pvt. Ltd. and contributors
# For license information, please see license.txt

"""SOP-005's supplier-approval check, run when a Purchase Order is submitted.

Warn-first by design: 943 of the site's suppliers have no approval status
yet, and hard-blocking every order on day one would stop purchasing dead.
A Suspended supplier always blocks -- suspension is a deliberate decision,
not a data gap. Flip SUPPLIER_GATE to "block" once the supplier base is
classified, and Not Approved suppliers block too.
"""

import frappe
from frappe import _

SUPPLIER_GATE = "warn"  # "block": Not Approved / unclassified suppliers also block


def check_supplier_approval(doc, method=None):
	status = frappe.db.get_value("Supplier", doc.supplier, "custom_approval_status")
	if status == "Approved":
		return
	if status == "Suspended":
		frappe.throw(
			_(
				"Supplier {0} is <b>Suspended</b> on the Approved Suppliers Register "
				"(REG-007). Orders cannot be placed until a new evaluation (FRM-005) "
				"restores approval."
			).format(frappe.bold(doc.supplier_name or doc.supplier)),
			title=_("Supplier Suspended"),
		)
	label = _(status) if status else _("not yet classified")
	message = _(
		"Supplier {0} is <b>{1}</b> on the Approved Suppliers Register (REG-007). "
		"SOP-005 expects purchases from Approved suppliers; record a Supplier "
		"Evaluation (FRM-005) to classify them."
	).format(frappe.bold(doc.supplier_name or doc.supplier), label)
	if SUPPLIER_GATE == "block":
		frappe.throw(message, title=_("Supplier Not Approved"))
	frappe.msgprint(message, title=_("Supplier Approval"), indicator="orange")
